Source/Components/PlayerComponent.cpp

```diff
@@ -354,26 +357,49 @@ void PlayerComponent::updateNote(int iNoteOnEventIndex, double fNewNoteOnTimesta
 void PlayerComponent::addNote(PianoRollNote * pPianoRollNote) {
     // Create midi messages
     auto noteOnMsg = MidiMessage::noteOn (1, pPianoRollNote->getNoteNumber(), (uint8) 120);
-    double a = convertQuarterNoteToSec(pPianoRollNote->getOffset());
-    noteOnMsg.setTimeStamp(convertQuarterNoteToSec(pPianoRollNote->getOffset()));
+    double noteOnTime = convertQuarterNoteToSec(pPianoRollNote->getOffset());
+    noteOnMsg.setTimeStamp(noteOnTime);
 
     auto noteOffMsg = MidiMessage::noteOff (1, pPianoRollNote->getNoteNumber(), (uint8) 120);
-    double d = convertQuarterNoteToSec(pPianoRollNote->getOffset() + pPianoRollNote->getLength());
-    noteOffMsg.setTimeStamp(d);
+    double noteOffTime = convertQuarterNoteToSec(pPianoRollNote->getOffset() + pPianoRollNote->getLength());
+    noteOffMsg.setTimeStamp(noteOffTime);
 
+    auto * pEventAtReadIdx = m_midiMessageSequence->getEventPointer(m_iMidiEventReadIdx);
+    
     // Add to midiMessageSequence
     auto * e1 = m_midiMessageSequence->addEvent(noteOnMsg);
     pPianoRollNote->setNoteOnEventPtr(e1);
     auto * e2 = m_midiMessageSequence->addEvent(noteOffMsg);
     pPianoRollNote->setNoteOffEventPtr(e2);
-
-    m_midiMessageSequence->updateMatchedPairs();
+    e1->noteOffObject = e2;
+
+    auto readIdxTime = m_midiMessageSequence->getEndTime();
+    if (pEventAtReadIdx)
+        readIdxTime = pEventAtReadIdx->message.getTimeStamp();
+    
+    if (noteOnTime <= readIdxTime)
+        m_iMidiEventReadIdx++;
+    if (noteOffTime <= readIdxTime)
+        m_iMidiEventReadIdx++;
+    
```

Comment:
Why is it necessary to remove updateMatchedPairs() ? The program seems to work correctly with updateMatchedPairs().

Reply:
Okay this snippet is efficient than using updateMatchedPairs() since it iterates through all the messages to fix missing noteOffs. 
I've moved this to a separate branch (https://github.com/Aavu/MIDI-Editor/tree/kaushal/update-notes-no-matched-pair-call) for now. We can merge it back here after I run some tests.

Reply:
I don't remember the details but I can paste my comments I once sent to the group chat:

I think I found the problem with updateNote.
It’s in MidiMessageSequence::updateMatchedPairs(). In that function it is creating new MidiEventHolder objects, this is why the old pointer does not work after updateMatchedPairs.
And for moving and changing the length of a note, we don’t need this function (the noteOn->noteOff pair is kept), just remove that line.